single-pool/js/src/transactions.ts

```diff
@@ -0,0 +1,302 @@
+  Connection,
+  Authorized,
+  Transaction,
+  StakeProgram,
+  SystemProgram,
+  StakeAuthorizationLayout,
+} from '@solana/web3.js';
+import {
+  MINT_SIZE,
+  getAssociatedTokenAddressSync,
+  createAssociatedTokenAccountInstruction,
+  createApproveInstruction,
+} from '@solana/spl-token';
+
+import {
+  findDefaultDepositAccountAddress,
+  findPoolAddress,
+  findPoolStakeAddress,
+  findPoolMintAddress,
+  findPoolStakeAuthorityAddress,
+  findPoolMintAuthorityAddress,
+} from './addresses';
+import { SinglePoolInstruction } from './instructions';
+
+interface DepositParams {
+  connection: Connection;
+  pool: PublicKey;
+  userWallet: PublicKey;
+  userStakeAccount?: PublicKey;
+  depositFromDefaultAccount?: boolean;
+  userTokenAccount?: PublicKey;
+  userLamportAccount?: PublicKey;
+  userWithdrawAuthority?: PublicKey;
+}
+
+interface WithdrawParams {
+  connection: Connection;
+  pool: PublicKey;
+  userWallet: PublicKey;
+  userStakeAccount: PublicKey;
+  tokenAmount: number | bigint;
+  createStakeAccount?: boolean;
+  userStakeAuthority?: PublicKey;
+  userTokenAccount?: PublicKey;
+  userTokenAuthority?: PublicKey;
+}
+
+export class SinglePoolProgram {
+  // FIXME replace with real id when we have one
+  static programId: PublicKey = new PublicKey('3cqnsMsT6LE96pxv7GR4di5rLqHDZZbR3FbeSUeRLFqY');
+
+  static space: number = 33;
+
+  static async initialize(
+    connection: Connection,
+    voteAccount: PublicKey,
+    payer: PublicKey,
+    skipMetadata = false,
+  ) {
+    const transaction = new Transaction();
+
+    const pool = findPoolAddress(this.programId, voteAccount);
+    const stake = findPoolStakeAddress(this.programId, pool);
+    const mint = findPoolMintAddress(this.programId, pool);
+
+    const poolRent = await connection.getMinimumBalanceForRentExemption(this.space);
+    const stakeRent = await connection.getMinimumBalanceForRentExemption(StakeProgram.space);
+    const mintRent = await connection.getMinimumBalanceForRentExemption(MINT_SIZE);
+    const minimumDelegation = (await connection.getStakeMinimumDelegation()).value;
+
+    transaction.add(
+      SystemProgram.transfer({
+        fromPubkey: payer,
+        toPubkey: pool,
+        lamports: poolRent,
+      }),
+    );
+
+    transaction.add(
+      SystemProgram.transfer({
+        fromPubkey: payer,
+        toPubkey: stake,
+        lamports: stakeRent + minimumDelegation,
+      }),
+    );
+
+    transaction.add(
+      SystemProgram.transfer({
+        fromPubkey: payer,
+        toPubkey: mint,
+        lamports: mintRent,
+      }),
+    );
+
+    transaction.add(SinglePoolInstruction.initializePool(voteAccount));
+
+    if (!skipMetadata) {
+      transaction.add(SinglePoolInstruction.createTokenMetadata(pool, payer));
+    }
+
+    return transaction;
+  }
+
+  static async deposit(params: DepositParams) {
+    const { connection, pool, userWallet } = params;
+
+    if (!params.userStakeAccount == !params.depositFromDefaultAccount) {
```

Comment:
I thought this was wrong for a bit, but then I read it some more and realized you're very clever -- nicely done!

Reply:
maybe too clever if its confusing! to me its just a logical xor

Reply:
Nah, I'm just a bit dense when it comes to xor checks